src/SonarQube.Client/Services/RequestFactory.cs

```diff
@@ -0,0 +1,98 @@
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace SonarQube.Client
+{
+    public class RequestFactory
+    {
+        private readonly Dictionary<Type, SortedList<Version, Func<object>>> requestMappings =
+            new Dictionary<Type, SortedList<Version, Func<object>>>();
+
+        /// <summary>
+        /// Registers a simple request factory for the specified version of SonarQube. It works only for types that
+        /// have parameterless constructor.
+        /// </summary>
+        /// <typeparam name="TRequest">The type of the request interface to create implementation for.</typeparam>
+        /// <typeparam name="TRequestImpl">The type implementing TRequest that has a parameterless constructor.</typeparam>
+        /// <param name="version">The version of SonarQube which first implements the request.</param>
+        /// <returns>Returns this.</returns>
+        public RequestFactory RegisterRequest<TRequest, TRequestImpl>(string version)
+            where TRequest : IRequest
+            where TRequestImpl : TRequest, new()
+        {
+            return RegisterRequest<TRequest, TRequestImpl>(version, () => new TRequestImpl());
+        }
+
+        /// <summary>
+        /// Registers a request factory for the specified version of SonarQube.
+        /// </summary>
+        /// <typeparam name="TRequest">The type of the request interface to create implementation for.</typeparam>
+        /// <typeparam name="TRequestImpl">The type implementing TRequest.</typeparam>
+        /// <param name="version">The version of SonarQube which first implements the request.</param>
+        /// <param name="factory">Factory function to create new instances of TRequestImpl.</param>
+        /// <returns>Returns this.</returns>
+        public RequestFactory RegisterRequest<TRequest, TRequestImpl>(string version, Func<TRequestImpl> factory)
+            where TRequest : IRequest
+            where TRequestImpl : TRequest
+        {
+            SortedList<Version, Func<object>> map;
+            if (!requestMappings.TryGetValue(typeof(TRequest), out map))
+            {
+                map = new SortedList<Version, Func<object>>();
+                requestMappings[typeof(TRequest)] = map;
+            }
+            map[Version.Parse(version)] = () => factory();
+            return this;
+        }
+
+        /// <summary>
+        /// Creates a new TRequest implementation for the specified SonarQube version.
+        /// </summary>
+        /// <typeparam name="TRequest">The type of the request implementation to create.</typeparam>
+        /// <param name="version">
+        /// SonarQube version to return a request implementation for. The default value returns the
+        /// latest registered implementation.QueryStringSerializer
```

Comment:
Should "QueryStringSerializer" be part of the comment? It's not clear why it's there.